### PR TITLE
Is_sorted, Is_sorted_until and Is_partitioned Algorithms

### PR DESCRIPTION
**Pull request by dcbdan**
Implemented the is_sorted, is_sorted_until and is_partitioned algorithms. 
I also fixed the formatting I had messed up on search.hpp.

### CODE REVIEW

**Comment by hkaiser** [comment]
@dcbdan: first of all, please rebase this branch on master so it can be merged later on.

File: docs/hpx.idx [hunk 1]
```diff
@@ -102,6 +102,13 @@ parallel::generate_n                  "generate_n" "hpx\.parallel\.v1\.generate_
 # hpx/parallel/algorithms/inclusive_scan.hpp
 parallel::inclusive_scan              "inclusive_scan" "hpx\.parallel\.v1\.inclusive_scan.*"
 
+# hpx/parallel/algorithms/is_partitioned.hpp
+parallel::is_partitioned              "is_partitioned" "hpx\.parallel\.v1\.is_partitioned.*"
+
+# hpx/parallel/algorithms/is_sorted.hpp
+parallel::is_sorted                   "is_sorted"  "hpx\.parallel\.v1\.is_sorted.*"
```
Comment (hkaiser) [inline docs/hpx.idx]: This regex will match `is_sorted_until` as well which will create a wrongly placed index entry.

File: hpx/parallel/algorithms/is_partitioned.hpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                        })
+                    );*/
+                    ////////////////////////////////////////////////////////////
+                    //    Using impl_is_partitioned()
+                    //To only call hpx::future::get() once for each iterator.
+                    //std::is_partitioned calls pred twice for the first element
+                    //that is false so [](hpx::future<bool> a) { return
+                    //a.get();} cannot be used as a predicate. Alternatively,
+                    //hpx::util::unwrapped can be used with std::is_partition.
+                    [tok](std::vector<hpx::future<bool> > && results) -> bool
```
Comment (hkaiser) [inline hpx/parallel/algorithms/is_partitioned.hpp]: Please remove comments if they are obsolete.

File: hpx/parallel/algorithms/is_sorted.hpp [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+                        util::loop_n(part_begin, part_size - 1,
+                            [&trail, &tok, &pred](FwdIter it)
+                            {
+                                if (pred(*it, *trail++))
+                                {
+                                    tok.cancel();
+                                }
+                            });
+                        FwdIter i = trail++;
+                        //trail now points one past the current grouping
```
Comment (hkaiser) [inline hpx/parallel/algorithms/is_sorted.hpp]: The definition of the variable `i` could be moved inside the block of the `if()` below.

File: hpx/parallel/algorithms/is_sorted.hpp [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+        typedef typename std::iterator_traits<FwdIter>::iterator_category
+            iterator_category;
+        typedef typename std::iterator_traits<FwdIter>::value_type
+            value_type;
+        BOOST_STATIC_ASSERT_MSG(
+                                (boost::is_base_of<
+                                 std::forward_iterator_tag, iterator_category
+                                 >::value),
+                                "Requires at least forward iterator.");
+
```
Comment (hkaiser) [inline hpx/parallel/algorithms/is_sorted.hpp]: The formatting of the static assert got messed up.

File: hpx/parallel/algorithms/is_sorted.hpp [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+                            [&trail, &tok, &pred](reference& v, std::size_t ind)
+                            {
+                                if (pred(v, *trail++))
+                                {
+                                    tok.cancel(ind);
+                                }
+                            });
+
+                        FwdIter i = trail++;
+
```
Comment (hkaiser) [inline hpx/parallel/algorithms/is_sorted.hpp]: Same as above, `I` can go into the block of the `if()` below.

File: hpx/parallel/algorithms/is_sorted.hpp (hunk 3 above)
Comment (hkaiser) [inline hpx/parallel/algorithms/is_sorted.hpp]: Sorry, I take that back.

**Comment by hkaiser** [comment]
I'm fine with merging this once the minor issues above have been addressed.

**Comment by dcbdan** [comment]
Did a rebase and made the changes. Of course, please make sure I didn't introduce anything incorrect.

**Pull request merged**